src/Evolution/Systems/GeneralizedHarmonic/BoundaryCorrections/UpwindPenalty.cpp

```diff
@@ -0,0 +1,292 @@
+        packaged_char_speed_n_times_v_minus,
+    const gsl::not_null<tnsr::aa<DataVector, Dim, Frame::Inertial>*>
+        packaged_char_speed_gamma2_v_spacetime_metric,
+    const gsl::not_null<tnsr::a<DataVector, 3, Frame::Inertial>*>
+        packaged_char_speeds,
+
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& spacetime_metric,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& pi,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& phi,
+
+    const Scalar<DataVector>& constraint_gamma1,
+    const Scalar<DataVector>& constraint_gamma2,
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& shift,
+
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_covector,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& normal_vector,
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+    /*mesh_velocity*/,
+    const std::optional<Scalar<DataVector>>& normal_dot_mesh_velocity)
+    const noexcept {
+  // Compute the char speeds without the mesh movement, then add the mesh
+  // movement. We compute the zero-speed first since it is just the normal
+  // dotted into the shift.
+  {
+    Scalar<DataVector> shift_dot_normal{};
+    get(shift_dot_normal)
+        .set_data_ref(make_not_null(&get<1>(*packaged_char_speeds)));
+    dot_product(make_not_null(&shift_dot_normal), shift, normal_covector);
+    get(shift_dot_normal) *= -1.0;
+    // the metric mode speed is the zero speed times (1 + gamma_1)
+    get<0>(*packaged_char_speeds) =
+        (1.0 + get(constraint_gamma1)) * get(shift_dot_normal);
+    // 2 = plus, 3 = minus
+    get<2>(*packaged_char_speeds) = get(lapse) + get(shift_dot_normal);
+    get<3>(*packaged_char_speeds) = -get(lapse) + get(shift_dot_normal);
+  }
+
+  if (normal_dot_mesh_velocity.has_value()) {
+    get<0>(*packaged_char_speeds) -= get(*normal_dot_mesh_velocity);
+    get<1>(*packaged_char_speeds) -= get(*normal_dot_mesh_velocity);
+    get<2>(*packaged_char_speeds) -= get(*normal_dot_mesh_velocity);
+    get<3>(*packaged_char_speeds) -= get(*normal_dot_mesh_velocity);
+  }
+
+  for (size_t a = 0; a < Dim + 1; ++a) {
+    for (size_t b = a; b < Dim + 1; ++b) {
+      packaged_char_speed_gamma2_v_spacetime_metric->get(a, b) =
+          get(constraint_gamma2) * spacetime_metric.get(a, b);
+    }
+  }
+
+  // Computes the contribution to the boundary correction from one side of the
+  // interface.
+  //
+  // Note: when UpwindPenalty::dg_boundary_terms() is called, an Element passes
+  // in its own packaged data to fill the interior fields, and its neighbor's
+  // packaged data to fill the exterior fields. This introduces a sign flip for
+  // each normal used in computing the exterior fields.
+  // Use v_psi allocation as n^i Phi_i
+  {
+    tnsr::aa<DataVector, Dim, Frame::Inertial>& normal_dot_phi =
+        *packaged_char_speed_v_spacetime_metric;
+    for (size_t a = 0; a < Dim + 1; ++a) {
```

Comment:
Yes, but I'm a bit hesitant to start using that until the design is completely settled. Not using it while it's under heavy development makes it a lot easier for people making major changes to TensorExpressions since there's little to no dependent code